### PR TITLE
Flatten OperationArguments and remove optional parameter local variables

### PR DESCRIPTION
**Pull request by daschult**
Resolves https://github.com/Azure/autorest.typescript/issues/188.

### CODE REVIEW

File: src/vanilla/Model/MethodTS.cs [hunk 1]
```diff
@@ -569,39 +532,28 @@ public string GenerateWithHttpOperationResponseMethod(string emptyLine)
             return builder.ToString();
         }
 
-        public void GenerateOperationArguments(TSArgumentList operationArguments)
+        public void GenerateOperationArguments(TSObject operationArguments)
         {
-            operationArguments.Object(obj =>
+            ParameterTransformations transformations = GetParameterTransformations();
+            Action<string, string> addArgument = (string operationArgumentName, string operationArgumentValue) =>
```
Comment (RikkiGibson) [inline src/vanilla/Model/MethodTS.cs]: Suggest using a local function for this

Comment (RikkiGibson) [reply src/vanilla/Model/MethodTS.cs]:
```cs
void addArgument(string operationArgumentName, string operationArgumentValue)
{
    ...
}
```

File: src/vanilla/Model/ParameterTransformations.cs [hunk 2]
```diff
@@ -13,6 +13,10 @@ namespace AutoRest.TypeScript.Model
     /// </summary>
     public class ParameterTransformations
     {
+        /// <summary>
+        /// The names of input parameters that are in the options parameter.
+        /// </summary>
+        private readonly ISet<string> inputParametersInOptions;
```
Comment (RikkiGibson) [inline src/vanilla/Model/ParameterTransformations.cs]: Does the parameter itself not know this based on `.Required`?

Comment (daschult) [reply src/vanilla/Model/ParameterTransformations.cs]: `.Required` doesn't quite cover whether or not a parameter is included in the `options` parameter. [This function](https://github.com/Azure/autorest.typescript/pull/211/files#diff-a0945a4f62f5a6e958b85ad35db7086eR756) was added to share the logic for what determines whether a parameter is in `options` or not.

File: test/azure/generated/AzureCompositeModelClient/operations/arrayModel.ts [hunk 3]
```diff
@@ -400,7 +401,10 @@ const putValidOperationSpec: msRest.OperationSpec = {
   ],
   requestBody: {
     parameterPath: {
-      arrayProperty: "arrayProperty"
+      arrayProperty: [
```
[the remark this answers is not in the document]
Comment (daschult) [reply test/azure/generated/AzureCompositeModelClient/operations/arrayModel.ts]: That's what I'm relying on.

File: src/vanilla/Model/ParameterTransformations.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/vanilla/Model/ParameterTransformations.cs]: Makes sense--things seem OK as is, then.

File: .gulp/gulpfile.iced [hunk 4]
```diff
@@ -62,13 +62,13 @@ task 'test/nodejs-unit', 'run nodejs unit tests', [], (done) ->
   await execute "#{basefolder}/node_modules/.bin/mocha", defer _
   done();
 
-task 'testci/chrome-unit', '', [], (done) ->
+task 'testci/chrome-unit', '', ['testci/nodejs-unit'], (done) ->
```
Comment (RikkiGibson) [inline .gulp/gulpfile.iced]: the CI jobs run the tests in parallel in different containers so I suggest not having a relationship between these tasks

File: .gulp/gulpfile.iced [hunk 5]
```diff
@@ -62,13 +62,13 @@ task 'test/nodejs-unit', 'run nodejs unit tests', [], (done) ->
...
 
-task 'testci/chrome-unit', '', [], (done) ->
+task 'testci/chrome-unit', '', ['testci/nodejs-unit'], (done) ->
   global.verbose = true
   await run "test/chrome-unit", defer _
   await execute "node #{basefolder}/.scripts/coverage", defer _
   done()
 
-task 'test/chrome-unit', 'run browser unit tests', [], (done) ->
+task 'test/chrome-unit', 'run browser unit tests', ['test/nodejs-unit'], (done) ->
```
Comment (RikkiGibson) [inline .gulp/gulpfile.iced]: Ordering these is probably for the best..

File: .gulp/gulpfile.iced (hunk 4 above)
Comment (daschult) [reply .gulp/gulpfile.iced]: Removed in https://github.com/Azure/autorest.typescript/pull/211/commits/65d262eca39c20e4affd29ec9f696aa810c2cf41.

File: src/vanilla/Model/MethodTS.cs (hunk 1 above)
Comment (daschult) [reply src/vanilla/Model/MethodTS.cs]: Done in https://github.com/Azure/autorest.typescript/pull/211/commits/65d262eca39c20e4affd29ec9f696aa810c2cf41.

**Comment by RikkiGibson** [comment]
Now that we run `npm install` without `--ignore-scripts` in the bundlesize check, a dotnet build occurs and it's necessary to modify .travis.yml to install the libunwind8 prerequisite.

You can either pass the `--ignore-scripts` flag or modify the .travis.yml to add the dotnet prerequisites:

```
    - env: CHECK_NAME="Bundle Size"
      install:
          # dotnet 2.0 prereq
        - sudo apt-get install libunwind8
        - npm install
      script: npm run bundle-size
```

This adds about a minute to the job.

File: .scripts/bundlesize.js [hunk 6]
```diff
@@ -23,26 +20,52 @@ function execVerbose(script) {
...
-  dependencies.refreshNodeModules();
-  const headSize = await getBundleSize();
-
-  const change = (headSize / baseSize) - 1;
-  const percentChange = (Math.abs(change) * 100).toFixed(2) + '%';
-
+/**
+ * Output a message to the console.
+ */
+function outputMessage(message) {
```
Comment (RikkiGibson) [inline .scripts/bundlesize.js]: Note that ci-sidecar will only consume the first `---output` block--subsequent blocks will be ignored

**Pull request merged**